### PR TITLE
[ECAM] Implement FWC flight phases and update ECAM messages

### LINKED ISSUE #942: [BUG] TO Config page appears if APU started after landing
**Mod Version**
Master

**Describe the bug**
After landing, if you start the APU while taxiing to the gate, the TO Config menu appears on the ECAM.

**To Reproduce**
1. Start a flight, take off, and land as normal.
2. While still taxiing to the gate, start the APU.

**Expected behavior**
The ECAM should remain in the same state as it was after touchdown.

**Actual behavior**
After the APU is started, the TO Config menu appears.

**Additional context**
I've noticed this happening on two flights in a row now.

### LINKED ISSUE #1067: [BUG] Landing page on Ecam reappears after landing
Master

The landing checklist on ecam reappears after the plane has already landed.

Land the plane, checklist will reappear on ECAM.

1. Approach phase
2. Landing Page checklist appears on Ecam
3. After plane has already landed page re appears while taxiing.

The page should not appear once the plane has already landed.

Page is displayed again.

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #1683: [BUG] T.O CONFIG AFTER LANDING
**0.4.1**

**Upon landing, when taxiing to a gate you get the ECAM MEMO for T.O configuration**

  ** landing, start APU and continue taxiing to gate
Check ECAM for TO MEMO**

1.After landing, start APU and continue taxiing to gate
2.Check ECAM for TO MEMO
3.

**There should not be a Take off configuration memo on the ECAM after landing**

**TO Config MEMO after landing when taxiing to gate**

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #1882: [BUG] Rarely, T.O cheklist remains displayed during all flight
**Mod Version**

Master 2020-11-10T15:19:38+00:00

**Describe the bug**

The take-off checklist remains displayed during all phases of flight

**To Reproduce**

1.  rarely, it is my first occurence
2.
3.

**Expected behavior**

check the conditions to displayed T.O checklist or to erase this message.

**Actual behavior**

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by beheh**
Fixes #942
Fixes #1067
Fixes #1683
Fixes #1882

## Summary of Changes

Per Asobo's default, the aircraft uses a unified, simple flight phase mechanism that is shared between all instruments (e.g. to differentiate takeoff vs. on ground behaviour). In some cases though, this is not detailled enough and the phases don't correlate with what the real aircraft does.

This pull request fully decouples the flight phases in the ECAM from the original flight phases, and implements the original FWC flight phase logic. This greatly enhances the accuracy of when certain ECAM messages appear, and in turn fixes some bugs like the T.O. memo appearing after landing. In future this should also simplify implementing new ECAM messages.

This PR doesn't fix nor implement every single ECAM message, but it's a stepping stone for more of that in future. It also enables a future PR to fully implement the ECAM message clear/recall/emerg cancel behaviour.

Detailed changelog for ECAM messages:
- OVERSPEED can no longer appear on the ground
- ENG FAIL can now appear on the ground in an appropriate flight phase
- BRAKES HOT no longer appears after landing until both engines are shut down for 5 minutes
- PARK BRK ON appears as an ECAM level 2 alert during flight (normal status only on ground)
- A/SKID N/WS OFF is inhibited during takeoff
- TCAS FAULT no longer appears during takeoff and landing
- IR IN ALIGN only appears before takeoff (amber during engine on not yet implemented) 
- LAND ASAP is inhibited during takeoff roll and landing roll
- T.O. INHIBIT and LDG INHIBIT show more appropriately now and have a 3 second delay
- SPEED BRK is inhibited after engine start and after landing
- AUTO BRK LO/MED/MAX is only shown when landing is imminent
- NW STRG DISC turns amber when an engine is running
- PRED W/S OFF turns amber during takeoff and landing
- COMPANY MSG is inhibited during takeoff and landing
- IGNITION added
- SIGNS...ON memo during takeoff memo now also requires NO SMOKING

## Screenshots (if necessary)

https://youtu.be/qUNgmQcwGOw (slightly older version - T.O and LDG INHIBIT have a 3s delay now according to documentation)

## References

grafik

image

image

### QA

For this one, watch the ECAM apart from the usual testing feel free to compare appearing messages various real sources.
While some messages are still missing as before, most of the ones that appear should only appear in an appropriate flight pase now. Various ECAM messages have also been subtly changed according to the docs, adjusting when they appear and in some cases how they appear (e.g. NW STRG DISC in amber while an engine is running).

Specifically If you encounter unexpected ECAM messages or an unexpected T.O/landing memos please take a note of the specific message, and describe how to reproduce it.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Benjozork** [comment]
Ate there any other instruments where we can reuse the same phase definitions ?

**Comment by beheh** [comment]
Hm, I don't think so. AFAIK the FMA is mostly driven by the FMC (which are mostly correct right now, but I have a little more cleanup planned there after #1886), and I can't think of anything else obvious that uses these.

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by beheh** [comment]
Keeping as draft for a bit while I'm investigating why the lower ecam runs the upper ecam code. Should not have a functional impact, it's just weird.

**Comment by beheh** [comment]
Waiting on #1965 to land.

**Comment by beheh** [comment]
Okay, #1965 is written, but I don't think that blocks the work here. Both PRs can land independently.

**Comment by beheh** [comment]
I feell ike I should move the FWC to A32NX_Core instead of the ECAM, so back to draft for a bit.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html:23]: I'm not super happy about importing thise here. An alternative would be to move the logic nodes to A32NX_Utils, but I wasn't sure about the convention and how it would be used.

**Comment by InhumanReactions** [comment]
Category: QA Tester
Name: InhumanReactions #7129
Date of testing: Current date 17/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1943
Object of test: [ECAM] Implement FWC flight phases and update ECAM messages

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Spawned in C&D at EGLL and set up the aircraft as standard.
2. Pushed back and the PRED W/S switched to being orange. Switched back to green after starting Engine #2 (not sure if this is supposed to happen).
3. Departed from EGLL. T/O Memo disappeared as expected when TOGA was activated. T.O Inhibit also activated with TOGA so all is working well. 
4. Flew the rest of the route up to EGGP
5. Landed normally, the Landing Memo came up at the correct time, and the Landing Inhibit memo also came up correctly.
6. Afterwards I loaded onto the runway, however the IRS wasn't aligned (for some unknown reason). The ECAM memos were working as listed. I attempted again, and the IRS worked this time and were aligned. Not sure if that is a problem or a one off. However both times the Signs switch wouldn't register movement of the switch and after being switched off, the ECAM displayed it as being on.

Flight notes (If required): Successful flight.

Issues (If Any): The Signs memo wouldn't register physical movement, even after flicking the switch multiple times. It also wouldn't register on landing, again I tried flicking the switch on and off. Even when starting on the runway, the changes still wouldn't register on the Upper ECAM. One last thing, the Lower ECAM display didn't switch from ENGINE to CRUISE throughout the flight. The phases in the MCDU were correct, but the ECAM stayed on ENGINE until I landed it switched to F/CTR and then subsequently APU when I started it up. Finally, it switched to the DOOR/OXY page and stayed there as usual. To test this load in C&D and go through a full flight. You may run into this issue too. After testing a go-around and selected heading, I can confirm the CRUISE page does work then. To reproduce it during a full flight, I would recommend starting C&D, as that is what I did and it didn't work.
Overall Rating: To be improved
Conclusions of the testing: This PR is really good aside from the issues mentioned above. Just the Lower ECAM not switching to CRUISE and the SIGNS memo not registering physical movement of the switch. Also the PRED W/S may not be fully working, but that may be how it is supposed to be, I am not sure. But yeah, good PR, just a few minor issues experienced. Good work! (Also sorry if this is too long, just a new tester and learning the most efficient ways to writing these up!)

**Comment by beheh** [comment]
Hey, thanks for taking my PR as your first QA session! Your test the feedback sound nice and detailed.

> Pushed back and the PRED W/S switched to being orange. Switched back to green after starting Engine #2 (not sure if this is supposed to happen).

That sounds like a bug. It should only turn amber once T.O. CONFIG is pressed (and then stay amber until takeoff). I'll look at it.

> the Lower ECAM display didn't switch from ENGINE to CRUISE 

I'm investigating this.

> However both times the Signs switch wouldn't register movement of the switch and after being switched off, the ECAM displayed it as being on.

As per the ESLD I've updated this memo item to require both the seatbelts to be on and the no smoking sign to be either on or auto. Is it possible you left the no smoking sign off (the default)?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html
Comment (nistei) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html:23]: The logic nodes seem to be incredibly versatile. I would suggest moving them to the Utils.

**Comment by beheh** [comment]
>Pushed back and the PRED W/S switched to being orange. Switched back to green after starting Engine #2 (not sure if this is supposed to happen).

While I'm unable to really reproduce this, the logic is a bit tricky and not quite correct the way it's right now. It should turn amber with at least one engine started and the T.O. Config button pressed, but that state and it's resets aren't fully mapped yet. For now I'm changing it so it's green until you're TOGA or in the air (at which point it's always amber, as per the docs).

> the Lower ECAM display didn't switch from ENGINE to CRUISE

I also can't reproduce this, and have a theory for this: You may have accidentally left your enginer starter switch near the pedestal on "IGN START". That forces the engine page. While I can't fix that, I've added the IGNITION message to the upper ECAM now as per the docs to remind you that you've left it on :-)

I believe that should address all the feedback from the test.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html
Comment (beheh) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html:23]: Sounds good. Any suggestion for a file name? I see the previous ones are in some kind of NX namespace - or is that somethnig else? How about `NXLogic.js`?

Comment (nistei) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html:23]: Sounds good to me.

Comment (beheh) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html]: Done.

**Comment by beheh** [comment]
I've gotten hold a more recent ESLD, so I've slightly tweaked ECAM message logic according to it. Would be good to get at least one more QA tester doinga full flight with the latest version.

**Comment by beheh** [comment]
I've rebased. On more QA Test would be good. I've also added a detailed changelog for the ECAM messages to the PR here itself (but too verbose for the actual CHANGELOG).

**Comment by CaptainR3X** [comment]
Category: QA Tester
Name: GustavRex#4347
Date of testing: Current date 19/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1943
Object of test: Implement FWC flight phases and update ECAM messages

Testing technique: Flight; EKYT-EKCH
Steps you've made to test the Pull Request:
1. C/D at EKYT, normal programming procedures, p/s ECAM performed normally
2. Taxi and lineup, tested that NW STRG DISC performed as it should, departure 
3. Cruise, checking ECAM functionality
4. Arrival programming, following same procedure as in the beginning
5. Appproach and short final (NRG STRG DISC, Com. msg and ECAM performance testing during landing), GA, approach & landing (same procedures as first attempt) taxi and shut-down

Flight notes (If required): Online, VatSim - multiple DIR and reprogramming, Go-Around 

Issues (If Any): None related to this PR but bug found in FCU (will post on discord)
Overall Rating: Good
Conclusions of the testing: Testing went fine. The ECAM performed as it should, even when trying to provoke failures. I tested many of the points from the changelog and was unable to find abnormal performances even after go-around. The ECAM was fully functional.

**Comment by wpine215** [comment]
This needs its own separate issue, but from looking at the ESLD, "IGNITION" shouldn't appear for CFM engines during normal operations, even when the engine selector switch is in ignition. Also, not sure about "IRS ALIGNED" message - perhaps a real pilot can confirm.

**Comment by beheh** [comment]
I'll get real pilot confirmation on the IRS ALIGNED - I went by this: 

image

...although looking at a reference more closely, it should a) disappear when either engine is running and b) according to a detailed reference it should say "IR ALIGNED", which makes more sense consistency-wise. I'll get confirmation and fix this in an upcoming ECAM PR.

Regarding CFM engines, I have this IGNITION memo reference here:

image

...but that refers to continuous ignition (and that's for all of PW/IAE/CFM). At the end of the day this would probably be implemented at that granularity in the FADEC, but I'll also check in with pilots if it appears during normal operations, then I'll also clean this up in an upcoming ECAM PR.

**Comment by beheh** [comment]
@wpine215 I've filed #2029 to track my work on the issues you raised.

**Pull request merged**
